crates/holochain_keystore/src/meta_lair_client.rs

```diff
@@ -1,44 +1,162 @@
+                loop {
+                    tokio::time::sleep(TIME_CHECK_FREQ).await;
+                    if c_check_send.send(()).is_err() {
+                        break;
+                    }
+                }
+            });
+        }
+
+        // setup the connection check logic
+        {
+            let inner = inner.clone();
+            let stub_tag: Arc<str> = CON_CHECK_STUB_TAG.to_string().into();
+            tokio::task::spawn(async move {
+                use tokio::sync::mpsc::error::TryRecvError;
+                'top_loop: while c_check_recv.recv().await.is_some() {
+                    'drain_queue: loop {
+                        match c_check_recv.try_recv() {
+                            Ok(_) => (),
+                            Err(TryRecvError::Empty) => break 'drain_queue,
+                            Err(TryRecvError::Disconnected) => break 'top_loop,
+                        }
+                    }
+
+                    let client = inner.lock().clone();
+
+                    // optimistic check - most often the stub will be there
+                    if client.get_entry(stub_tag.clone()).await.is_ok() {
+                        continue;
+                    }
+
+                    // on the first run of a new install we need to create
+                    let _ = client.new_seed(stub_tag.clone(), None, false).await;
+
+                    // then we can exit early again
+                    if client.get_entry(stub_tag.clone()).await.is_ok() {
+                        continue;
+                    }
+
+                    // we couldn't fetch the stub, enter our reconnect loop
+                    let mut backoff_ms = RECON_INIT_MS;
+                    'reconnect: loop {
+                        'drain_queue2: loop {
+                            match c_check_recv.try_recv() {
+                                Ok(_) => (),
+                                Err(TryRecvError::Empty) => break 'drain_queue2,
+                                Err(TryRecvError::Disconnected) => break 'top_loop,
+                            }
+                        }
+
+                        backoff_ms *= 2;
+                        if backoff_ms >= RECON_MAX_MS {
+                            backoff_ms = RECON_MAX_MS;
+                        }
+                        tokio::time::sleep(std::time::Duration::from_millis(backoff_ms)).await;
+                        let opts = IpcKeystoreClientOptions {
+                            connection_url: connection_url.clone().into(),
+                            passphrase: passphrase.clone(),
+                            exact_client_server_version_match: true,
+                        };
+
+                        tracing::warn!("lair connection lost, attempting reconnect");
+
+                        let client = match ipc_keystore_connect_options(opts).await {
+                            Err(err) => {
+                                tracing::error!(?err, "lair connect error");
```

Comment:
Can one of the errors simply be "couldn't make connection?" That doesn't seem worth an error log. Maybe both of these can be warn level or lower?